jenkins/Jenkinsfile

```diff
@@ -0,0 +1,240 @@
+
+  def filteredImages = []
+  //we must use a c-style loop due to jenkins serialization, aka https://www.devopscat.tech/2018/10/what-is-noncps-in-jenkins-pipeline/
+  for (String currImage in images) {
+    if (hasImageChanged(currImage)) {
+      println("detected change for $currImage")
+      filteredImages.add(currImage) 
+    }
+  }
+
+  return filteredImages
+}
+
+def Boolean hasImageChanged(dir) {
+  def String output = sh (
+        script: "git diff HEAD..HEAD~1 $dir",
+        returnStdout: true
+  ).trim()
+
+  !output.isEmpty()
+}
+
+def void buildAll(ArrayList<String> imageDirs) {
+  for (String imageDir in imageDirs) {
+    buildImage(imageDir)
+  }
+}
+
+def void buildImage(String imageDir) {
+  def workingDir = pwd()
+  sh "echo 'building the following image: $imageDir'"
+  def exitCode = sh(
+    script: "$workingDir/build.sh $imageDir",
+    returnStatus: true
+  )
+
+  //if we cannot build an image because it will cause an overwrite of the remote gcr image, we set the build to unstable. 
+  //We do this because the most likely scenario is it was desired that the image to build, but the VERSION file was not updated.
+  //we do not want to fail in this circumstance, as there are other images to build
+  if (exitCode == 14) {
+    println("The build script reported that building would result in overwritting an existing image. Setting build status to UNSTABLE. Consider updating the VERSION file for $imageDir")
+    currentBuild.setResult("UNSTABLE")
+    prevDesc = currentBuild.getDescription() == null ? "" : currentBuild.getDescription() + "\n"
+    currentBuild.setDescription(prevDesc + "The build for image $imageDir failed because the image already exists.")
+  } else if (exitCode != 0) {
+    throw new RuntimeException("Unexpected error code recieved from build.sh: $exitCode")
+  }
+  //we do nothing in the else case, the shell command exitted as expected
+}
+
+def LinkedHashMap<String, String> getVersionMap(imageDirs) {
+  def LinkedHashMap<String, String>  versionMap = [:]
+
+  for (String currImageDir in imageDirs) {
+    def workingDir = pwd()
+    def version = sh(
+      script: "cat $workingDir/$currImageDir/VERSION",
+      returnStdout: true
+    ).trim()
+    versionMap[currImageDir] = version
+  }
+
+  versionMap
+}
+
+//versionMap must be defined by a previous step for this to work
+def void updateConfigHashes(replacementMap) {
+  def String[] configPaths = readFile("jenkins/confLocations.txt").trim().split("\n")
+  for (String currConfig in configPaths) {
+    println("Calling replaceTag function for $currConfig with $replacementMap")
+    replaceTags(currConfig, replacementMap)
+
+    println("printing changes in $currConfig and running git add")
+    sh(
+      script: "git diff $currConfig; git add $currConfig"
+    )
+  }
+}
+
+//parameterize with ":" prefix and end quote suffix instead of hardcode?
+def void replaceTags(String filePath, LinkedHashMap<String, String> replacementMap) {
```

Comment:
can you add comment that has an example of what the tag looks like before and after this function?